File: windows/tests/Volar.Domain.Tests/TaskSectionsTests.cs
using Volar.Core;
using Volar.Domain;
using Xunit;

namespace Volar.Domain.Tests;

public class TaskSectionsTests
{
    // Fixed, non-DST zone (UTC+7); no test reads the system clock.
    private static readonly TimeZoneInfo Ict = TimeZoneInfo.CreateCustomTimeZone("ICT-sections-test", TimeSpan.FromHours(7), "ICT", "ICT");

    // Monday, 2026-03-16, 09:00 +07:00.
    private static readonly DateTimeOffset Now = new(2026, 3, 16, 9, 0, 0, TimeSpan.FromHours(7));

    private static DateTimeOffset Cut => TaskSections.StartOfTomorrow(Now, Ict);

    private static DateTimeOffset Local(int day, int hour) => new(2026, 3, day, hour, 0, 0, TimeSpan.FromHours(7));

    private static TaskItem Task(
        DateTimeOffset? deadline = null,
        IReadOnlyList<Condition>? conditions = null,
        TaskState status = TaskState.Todo) =>
        new()
        {
            Id = Guid.NewGuid(),
            Title = "t",
            Details = string.Empty,
            Status = status,
            Deadline = deadline,
            Conditions = conditions ?? Array.Empty<Condition>(),
            CreatedAt = Now,
        };

    // MARK: - The day boundary itself

    [Fact]
    public void StartOfTomorrow_IsLocalMidnight_NotNowPlus24h()
    {
        var local = TimeZoneInfo.ConvertTime(Cut, Ict);
        Assert.Equal(new DateTime(2026, 3, 17, 0, 0, 0), local.DateTime);
    }

    // MARK: - Inbox

    [Fact]
    public void Inbox_NoDeadlineNoConditions_IsInbox()
    {
        Assert.True(TaskSections.IsInbox(Task()));
    }

    [Fact]
    public void Inbox_WithDeadline_IsNotInbox()
    {
        Assert.False(TaskSections.IsInbox(Task(deadline: Local(20, 9))));
    }

    [Fact]
    public void Inbox_WithAnyCondition_IsNotInbox()
    {
        // Blocked-on-another-task is a chain, not an unsorted capture — even though it has no date.
        var blocked = Task(conditions: new Condition[] { new TaskDoneCondition(Guid.NewGuid()) });
        Assert.False(TaskSections.IsInbox(blocked));
    }

    [Theory]
    [InlineData(TaskState.Done)]
    [InlineData(TaskState.Archived)]
    public void Inbox_ResolvedTask_IsNotInbox(TaskState status)
    {
        Assert.False(TaskSections.IsInbox(Task(status: status)));
    }

    // MARK: - Upcoming: deadlines

    [Fact]
    public void Upcoming_DeadlineTomorrow_IsUpcoming()
    {
        Assert.Equal(Local(17, 9), TaskSections.UpcomingDate(Task(deadline: Local(17, 9)), Cut));
    }

    [Fact]
    public void Upcoming_DeadlineLaterToday_IsNotUpcoming()
    {
        // Still due today -> Today owns it, even though the instant is in the future.
        Assert.Null(TaskSections.UpcomingDate(Task(deadline: Local(16, 23)), Cut));
    }

    [Fact]
    public void Upcoming_OverdueDeadline_IsNotUpcoming()
    {
        Assert.Null(TaskSections.UpcomingDate(Task(deadline: Local(14, 9)), Cut));
    }

    [Fact]
    public void Upcoming_DeadlineExactlyAtMidnight_IsUpcoming()
    {
        // The boundary is inclusive on the Upcoming side: 00:00 tomorrow is a later calendar day.
        Assert.Equal(Cut, TaskSections.UpcomingDate(Task(deadline: Cut), Cut));
    }

    // MARK: - Upcoming: deferred tasks (the reason this section exists at all)

    [Fact]
    public void Upcoming_DeferredToFutureDate_IsUpcoming_EvenWithNoDeadline()
    {
        var deferred = Task(conditions: new Condition[] { new AfterDateCondition(Local(19, 8)) });
        Assert.Equal(Local(19, 8), TaskSections.UpcomingDate(deferred, Cut));
    }

    [Fact]
    public void Upcoming_DeferPassed_IsNotUpcoming()
    {
        var deferred = Task(conditions: new Condition[] { new AfterDateCondition(Local(15, 8)) });
        Assert.Null(TaskSections.UpcomingDate(deferred, Cut));
    }

    [Fact]
    public void Upcoming_NonDateCondition_DoesNotQualify()
    {
        var blocked = Task(conditions: new Condition[] { new ExternalCondition("waiting on Bob", false) });
        Assert.Null(TaskSections.UpcomingDate(blocked, Cut));
    }

    [Fact]
    public void Upcoming_DeadlineAndDefer_GroupsUnderTheEarlier()
    {
        // Deferred to Wednesday, due Friday: Wednesday is the next date the user must care about.
        var both = Task(
            deadline: Local(20, 17),
            conditions: new Condition[] { new AfterDateCondition(Local(18, 8)) });
        Assert.Equal(Local(18, 8), TaskSections.UpcomingDate(both, Cut));
    }

    [Fact]
    public void Upcoming_FutureDeadlineWithPastDefer_UsesTheDeadline()
    {
        var both = Task(
            deadline: Local(20, 17),
            conditions: new Condition[] { new AfterDateCondition(Local(15, 8)) });
        Assert.Equal(Local(20, 17), TaskSections.UpcomingDate(both, Cut));
    }

    [Fact]
    public void Upcoming_MultipleFutureDefers_UsesTheEarliest()
    {
        var many = Task(conditions: new Condition[]
        {
            new AfterDateCondition(Local(22, 8)),
            new AfterDateCondition(Local(18, 8)),
            new AfterDateCondition(Local(25, 8)),
        });
        Assert.Equal(Local(18, 8), TaskSections.UpcomingDate(many, Cut));
    }

    [Theory]
    [InlineData(TaskState.Done)]
    [InlineData(TaskState.Archived)]
    public void Upcoming_ResolvedTask_IsNotUpcoming(TaskState status)
    {
        Assert.Null(TaskSections.UpcomingDate(Task(deadline: Local(20, 9), status: status), Cut));
    }

    // MARK: - The sections must not overlap

    [Fact]
    public void InboxAndUpcoming_AreDisjoint()
    {
        var candidates = new[]
        {
            Task(),
            Task(deadline: Local(20, 9)),
            Task(deadline: Local(16, 23)),
            Task(conditions: new Condition[] { new AfterDateCondition(Local(19, 8)) }),
            Task(conditions: new Condition[] { new TaskDoneCondition(Guid.NewGuid()) }),
        };

        foreach (var task in candidates)
        {
            Assert.False(TaskSections.IsInbox(task) && TaskSections.IsUpcoming(task, Cut));
        }
    }
}

File: Volar/Tests/TaskSectionsTests.swift
// Tests/TaskSectionsTests.swift — XCTest coverage for Sources/Model/TaskSections.swift, the
// sidebar Today/Upcoming/Inbox membership rules. Mirrors the Windows reference suite's case list
// 1:1 (voci-windows/windows/tests/Volar.Domain.Tests/TaskSectionsTests.cs), adapted to
// TaskItem/VolarCore.Condition's Swift shapes instead of C#'s Task/Condition.
//
// UNVERIFIED (written entirely on Windows — no Xcode/xcodegen/simulator available here): confirm on
// Mac that `xcodegen generate` + `xcodebuild test` actually discovers and runs this file under the
// `VolarTests` target, same as every other file already in this directory.
import Foundation
import XCTest
import VolarCore
@testable import Volar

final class TaskSectionsTests: XCTestCase {

    // MARK: - Deterministic time fixtures (fixed, non-DST zone; nothing here reads the system clock)

    private static let ict = TimeZone(identifier: "Asia/Bangkok")! // UTC+7, no DST — mirrors the
    // Windows suite's own fixed "ICT-sections-test" custom zone.

    private static var calendar: Calendar {
        var calendar = Calendar(identifier: .gregorian)
        calendar.timeZone = ict
        return calendar
    }

    /// Monday 2026-03-16, 09:00 +07:00 — same reference instant as the Windows suite.
    private static let now: Date = {
        var components = DateComponents()
        components.year = 2026
        components.month = 3
        components.day = 16
        components.hour = 9
        components.minute = 0
        components.second = 0
        return calendar.date(from: components)!
    }()

    private static var cutoff: Date { TaskSections.startOfTomorrow(now: now, timeZone: ict) }

    /// A local instant on 2026-03-`day` at `hour`:00 +07:00.
    private static func local(_ day: Int, _ hour: Int) -> Date {
        var components = DateComponents()
        components.year = 2026
        components.month = 3
        components.day = day
        components.hour = hour
        components.minute = 0
        components.second = 0
        return calendar.date(from: components)!
    }

    private func makeTask(
        deadline: Date? = nil,
        conditions: [VolarCore.Condition] = [],
        status: TaskStatus = .todo
    ) -> TaskItem {
        TaskItem(title: "t", priority: .medium, status: status, deadline: deadline, conditions: conditions, when: .later)
    }

    // MARK: - The day boundary itself

    func testStartOfTomorrowIsLocalMidnightNotNowPlus24h() {
        let components = Self.calendar.dateComponents(
            [.year, .month, .day, .hour, .minute, .second],
            from: Self.cutoff
        )
        XCTAssertEqual(components.year, 2026)
        XCTAssertEqual(components.month, 3)
        XCTAssertEqual(components.day, 17)
        XCTAssertEqual(components.hour, 0)
        XCTAssertEqual(components.minute, 0)
        XCTAssertEqual(components.second, 0)
    }

    // MARK: - Inbox

    func testInboxNoDeadlineNoConditionsIsInbox() {
        XCTAssertTrue(TaskSections.isInbox(makeTask()))
    }

    func testInboxWithDeadlineIsNotInbox() {
        XCTAssertFalse(TaskSections.isInbox(makeTask(deadline: Self.local(20, 9))))
    }

    func testInboxWithAnyConditionIsNotInbox() {
        // Blocked-on-another-task is a chain, not an unsorted capture — even though it has no date.
        let blocked = makeTask(conditions: [.taskDone(UUID())])
        XCTAssertFalse(TaskSections.isInbox(blocked))
    }

    func testInboxResolvedTaskIsNotInbox() {
        XCTAssertFalse(TaskSections.isInbox(makeTask(status: .done)))
        XCTAssertFalse(TaskSections.isInbox(makeTask(status: .archived)))
    }

    // MARK: - Upcoming: deadlines

    func testUpcomingDeadlineTomorrowIsUpcoming() {
        let task = makeTask(deadline: Self.local(17, 9))
        XCTAssertEqual(TaskSections.upcomingDate(task, startOfTomorrow: Self.cutoff), Self.local(17, 9))
    }

    func testUpcomingDeadlineLaterTodayIsNotUpcoming() {
        // Still due today -> Today owns it, even though the instant is in the future.
        let task = makeTask(deadline: Self.local(16, 23))
        XCTAssertNil(TaskSections.upcomingDate(task, startOfTomorrow: Self.cutoff))
    }

    func testUpcomingOverdueDeadlineIsNotUpcoming() {
        let task = makeTask(deadline: Self.local(14, 9))
        XCTAssertNil(TaskSections.upcomingDate(task, startOfTomorrow: Self.cutoff))
    }

    func testUpcomingDeadlineExactlyAtMidnightIsUpcoming() {
        // The boundary is inclusive on the Upcoming side: 00:00 tomorrow is a later calendar day.
        let task = makeTask(deadline: Self.cutoff)
        XCTAssertEqual(TaskSections.upcomingDate(task, startOfTomorrow: Self.cutoff), Self.cutoff)
    }

    // MARK: - Upcoming: deferred tasks (the reason this section exists at all)

    func testUpcomingDeferredToFutureDateIsUpcomingEvenWithNoDeadline() {
        let deferred = makeTask(conditions: [.afterDate(Self.local(19, 8))])
        XCTAssertEqual(TaskSections.upcomingDate(deferred, startOfTomorrow: Self.cutoff), Self.local(19, 8))
    }

    func testUpcomingDeferPassedIsNotUpcoming() {
        let deferred = makeTask(conditions: [.afterDate(Self.local(15, 8))])
        XCTAssertNil(TaskSections.upcomingDate(deferred, startOfTomorrow: Self.cutoff))
    }

    func testUpcomingNonDateConditionDoesNotQualify() {
        let blocked = makeTask(conditions: [.external(description: "waiting on Bob", satisfied: false)])
        XCTAssertNil(TaskSections.upcomingDate(blocked, startOfTomorrow: Self.cutoff))
    }

    func testUpcomingDeadlineAndDeferGroupsUnderTheEarlier() {
        // Deferred to Wednesday, due Friday: Wednesday is the next date the user must care about.
        let both = makeTask(deadline: Self.local(20, 17), conditions: [.afterDate(Self.local(18, 8))])
        XCTAssertEqual(TaskSections.upcomingDate(both, startOfTomorrow: Self.cutoff), Self.local(18, 8))
    }

    func testUpcomingFutureDeadlineWithPastDeferUsesTheDeadline() {
        let both = makeTask(deadline: Self.local(20, 17), conditions: [.afterDate(Self.local(15, 8))])
        XCTAssertEqual(TaskSections.upcomingDate(both, startOfTomorrow: Self.cutoff), Self.local(20, 17))
    }

    func testUpcomingMultipleFutureDefersUsesTheEarliest() {
        let many = makeTask(conditions: [
            .afterDate(Self.local(22, 8)),
            .afterDate(Self.local(18, 8)),
            .afterDate(Self.local(25, 8)),
        ])
        XCTAssertEqual(TaskSections.upcomingDate(many, startOfTomorrow: Self.cutoff), Self.local(18, 8))
    }

    func testUpcomingResolvedTaskIsNotUpcoming() {
        let done = makeTask(deadline: Self.local(20, 9), status: .done)
        let archived = makeTask(deadline: Self.local(20, 9), status: .archived)
        XCTAssertNil(TaskSections.upcomingDate(done, startOfTomorrow: Self.cutoff))
        XCTAssertNil(TaskSections.upcomingDate(archived, startOfTomorrow: Self.cutoff))
    }

    // MARK: - The sections must not overlap

    func testInboxAndUpcomingAreDisjoint() {
        let candidates: [TaskItem] = [
            makeTask(),
            makeTask(deadline: Self.local(20, 9)),
            makeTask(deadline: Self.local(16, 23)),
            makeTask(conditions: [.afterDate(Self.local(19, 8))]),
            makeTask(conditions: [.taskDone(UUID())]),
        ]

        for task in candidates {
            XCTAssertFalse(TaskSections.isInbox(task) && TaskSections.isUpcoming(task, startOfTomorrow: Self.cutoff))
        }
    }
}
